license-details/README.md

```diff
@@ -0,0 +1,138 @@
+# Introduction
+
+The P4 compiler `p4c` is a leading project of the P4 Consortium, and
+used by several companies as the front end of P4 compilers for a
+variety of hardware targets.  Most of these companies develop
+significant proprietary back end code, and the resulting binary
+executables are released, with no copyleft obligation for those
+companies to release their back end source code.
+
+This is allowed, because the `p4c` front source files are licensed
+under the Apache 2.0 license.  This license was chosen for multiple
+reasons, but in part to enable companies to do this.  It is the P4
+Consortium's intent to continue enabling this.
+
+
+# Linux Foundation review of P4.org software licenses in published code
+
+In early 2024 P4.org became part of The Linux Foundation.  In November
+2024 the Linux Foundation appointed someone to do a quick review of
+software licenses in the public repositories in the
+https://github.com/p4lang organization.
+
+They recommended a fairly light-weight mechanism to make the software
+license of all source files explicit and easy to track with automated
+tools: adding an SPDX-License-Identifier comment line [1].
+
+[1] https://spdx.org/licenses/
+
+These look like the following examples, where the syntax of everything
+_except_ the comment indicator is required by SPDX, i.e. there is a
+precise specified syntax for ethe "SPDX-License-Identifier:"
+characters to be present, and exactly what sequence of characters may
+follow it on the line.
+
+```c++
+// SPDX-License-Identifier: Apache-2.0
+```
+
+```python
+# SPDX-License-Identifier: GPL-2.0-only
+```
+
+
+# Why is the top level LICENSE file insufficient?
+
+One short answer is that not all files in this repository can legally
+be released under the Apache 2.0 license.  Source files that are an
+exception are described below.  We expect them to be a fairly small
+set of files, and that these exceptions are _not_ among the source
+files that are compiled to create a P4 compiler executable binary.
+
+Given that, a per-file software license tracking mechanism becomes
+useful.  It is also useful when individual files or subsets of files
+are copied into other open source software projects.
+
+
+# Most source files will be released with Apache 2.0 license
+
+The Apache 2.0 license is the default choice for all source files in
+https://github.com/p4lang projects, unless this is not legally
+allowed.
+
+
+# Source files that cannot be released under the Apache 2.0 license
+
+## Python source files that import the Scapy library
```

Comment:
We should probably get rid of scapy entirely if it causes such head aches. We practically only use it for a couple function calls.

Reply:
I think it is more than a couple.  grep through the p4c repo for occurrences of scapy, and you can see that in the ebpf back end test code, it uses a fair number of functions, not just `Ether`.

Are you mainly hoping to eliminate the use of scapy in the p4c repo only?  Or also the following p4lang repos?
+ tutorials
+ switch
+ p4-dpdk-target
+ behavioral-model
+ p4app
+ ptf

Reply:
Also I suspect that ptf Python code (most of the ptf code) probably needs to be GPL v2 because it imports scapy, and p4c imports ptf in several places, too.

Reply:
At least for eBPF I remember writing this code and it was mostly used for sniffing packets. There was new code added for uBPF however which we would need to take a look at. 

PTF already has been changed to be independent of scapy with https://github.com/p4lang/ptf/pull/147. We just need to add a module and switch the default. Could be a GSoC project :).

Reply:
It is also important that the compiler code itself has no GPL dependencies. Just the test scaffolding we use may involve GPL components.

Reply:
> It is also important that the compiler code itself has no GPL dependencies. Just the test scaffolding we use may involve GPL components.

If the current PR does not make exactly this point very clear, please tell me how to make it more clear.

Reply:
> If the current PR does not make exactly this point very clear, please tell me how to make it more clear.

For example, the section of line 44 gives the impression that these are compiler source files, when they are in fact test framework source files. There is no explicit mention that both scapy and eBPF dependencies are unrelated to compiler source code.

Reply:
In commit 4, I have added a new short first paragraph explicitly mentioning this to the section you mention.  I also added a note to the beginning of the next section, but it will need updating after we confirm the result of a more careful job.